src/createStreamMiddleware.ts

```diff
@@ -129,8 +130,15 @@ export default function createStreamMiddleware(options: Options = {}) {
    * Retry pending requests.
    */
   function retryStuckRequests() {
-    Object.values(idMap).forEach(({ req }) => {
-      // TODO: limiting retries could be implemented here
+    Object.values(idMap).forEach(({ req, retryCount = 0 }) => {
+      // Avoid retrying requests without an id - they cannot have matching responses so retry logic doesn't apply
+      // Check for retry count below ensure that a request is not retried more than 3 times
+      if (!req.id || retryCount >= 3) {
+        throw new Error(
+          `StreamMiddleware - Retry limit exceeded for request id "${req.id}"`,
```

Comment:
Hmm. This would be a confusing error message if `req.id` was not set.

Perhaps we can avoid setting requests with no id on `idMap` in the first place instead? Given that they are keyed by id, it doesn't really make sense to include them.

Reply:
The code to set request into idMap is old, I would avoid changing it.

Reply:
Do you see what I mean about this error though? If `req.id` is `undefined` here, we're going to get a nonsensical error about the retry limit being exceeded for id `undefined`. But in fact there was no retry limit exceeded.

I don't think it matters how old the code is that we're changing; if it's not suited for what we're doing today, we should change it. I'm just noticing now that this mistake of putting a request with no id in this map is also responsible for the type mangling on like 56; so what we have now isn't even type safe.

But regardless what we do with `idMap`, I don't think we should throw such a confusing error in this circumstance.

Reply:
I agree to your point about error, I updated it.

I noticed requests with id `undefined` being sent atm, thus I would avoid changing in this PR. It needs more work and also testing.

Reply:
Thanks!